src/main/java/io/supertokens/oauth/OAuth.java

```diff
@@ -66,39 +101,181 @@ public static OAuthAuthResponse getAuthorizationUrl(Main main, AppIdentifier app
+                accessToken = reSignToken(appIdentifier, main, accessToken, iss, 1, useDynamicKey, 0);
+                response.addProperty("access_token", accessToken);
+            }
+
+            if (response.has("id_token")) {
+                String idToken = response.get("id_token").getAsString();
+                idToken = reSignToken(appIdentifier, main, idToken, iss, 2, useDynamicKey, 0);
+                response.addProperty("id_token", idToken);
+            }
+
+            if (response.has("refresh_token")) {
+                String refreshToken = response.get("refresh_token").getAsString();
+                refreshToken = refreshToken.replace("ory_rt_", "st_rt_");
+                response.addProperty("refresh_token", refreshToken);
+            }
+            return response;
+
+        } catch (HttpResponseException e) {
+            JsonObject errorResponse = new Gson().fromJson(e.rawMessage, JsonObject.class);
+            throw new OAuthAPIException(
+                errorResponse.get("error").getAsString(),
+                errorResponse.get("error_description").getAsString(),
+                e.statusCode
+            );
+        }
+    }
+
+    private static String reSignToken(AppIdentifier appIdentifier, Main main, String token, String iss, int stt, boolean useDynamicSigningKey, int retryCount) throws IOException, JWTException, InvalidKeyException, NoSuchAlgorithmException, StorageQueryException, StorageTransactionLogicException, UnsupportedJWTSigningAlgorithmException, TenantOrAppNotFoundException, InvalidKeySpecException, JWTCreationException, InvalidConfigException, OAuthAPIException {
+        // Load the JWKS from the specified URL
+        String publicOAuthProviderServiceUrl = Config.getConfig(appIdentifier.getAsPublicTenantIdentifier(), main).getOAuthProviderPublicServiceUrl();
+        String jwksUrl = publicOAuthProviderServiceUrl + HYDRA_JWKS_PATH; // Use the new constant
+
+        // Check if cached keys are available for the jwksUrl
+        Map<String, JsonObject> cachedKeys = jwksCache.get(jwksUrl);
+        if (cachedKeys == null) {
+            JsonObject jwksResponse;
+            try {
+                jwksResponse = HttpRequest.sendGETRequest(main, "", jwksUrl, null, 10000, 10000, null);
+            } catch (HttpResponseException e) {
+                throw new OAuthAPIException("jwks_error", "Could not fetch JWKS keys from hydra for token verification", 500);
+            }
+            cachedKeys = new HashMap<>();
+            JsonArray keysArray = jwksResponse.get("keys").getAsJsonArray();
+
+            // Populate the cache with keys indexed by kid
+            for (JsonElement key : keysArray) {
+                JsonObject keyObject = key.getAsJsonObject();
+                String kid = keyObject.get("kid").getAsString();
+                cachedKeys.put(kid, keyObject);
+            }
+            jwksCache.put(jwksUrl, cachedKeys); // Cache the keys with jwksUrl as the key
+        }
```

Comment:
I feel like the HydraJWKCache could be moved into a separate class/file, but I'm not sure if we actually need to validate the tokens here.
They come as a response to a request we make (and we query the same server for the keys we use for verification). There may be some upside to it, but I don't see it right now.

Reply:
JWT verification is removed.